library/core/src/macros/mod.rs

```diff
@@ -284,6 +285,95 @@ pub macro cfg_match {
     }
 }
 
+/// A macro for defining `#[cfg]` match-like statements.
+///
+/// It is similar to the `if/elif` C preprocessor macro by allowing definition of a cascade of
+/// `#[cfg]` cases, emitting the implementation which matches first.
+///
+/// This allows you to conveniently provide a long list `#[cfg]`'d blocks of code
+/// without having to rewrite each clause multiple times.
+///
+/// Trailing `_` wildcard match arms are **optional** and they indicate a fallback branch when
+/// all previous declarations do not evaluate to true.
+///
+/// # Example
+///
+/// ```
+/// #![feature(cfg_match)]
+///
+/// cfg_match! {
+///     (unix) => {
+///         fn foo() { /* unix specific functionality */ }
+///     }
+///     (target_pointer_width = "32") => {
+///         fn foo() { /* non-unix, 32-bit functionality */ }
+///     }
+///     _ => {
+///         fn foo() { /* fallback implementation */ }
+///     }
+/// }
+/// ```
+#[cfg(not(bootstrap))]
+#[unstable(feature = "cfg_match", issue = "115585")]
+#[rustc_diagnostic_item = "cfg_match"]
+pub macro cfg_match {
+    // with a final wildcard
+    (
+        $(($initial_meta:meta) => { $($initial_tokens:tt)* })+
+        _ => { $($extra_tokens:tt)* }
+    ) => {
+        cfg_match! {
+            @__items ();
+            $((($initial_meta) ($($initial_tokens)*)),)+
+            (() ($($extra_tokens)*)),
+        }
+    },
+
+    // without a final wildcard
+    (
+        $(($extra_meta:meta) => { $($extra_tokens:tt)* })*
+    ) => {
+        cfg_match! {
+            @__items ();
+            $((($extra_meta) ($($extra_tokens)*)),)*
+        }
+    },
+
+    // Internal and recursive macro to emit all the items
+    //
+    // Collects all the previous cfgs in a list at the beginning, so they can be
+    // negated. After the semicolon is all the remaining items.
+    (@__items ($($_:meta,)*);) => {},
+    (
+        @__items ($($no:meta,)*);
+        (($($yes:meta)?) ($($tokens:tt)*)),
+        $($rest:tt,)*
+    ) => {
+        // Emit all items within one block, applying an appropriate #[cfg]. The
+        // #[cfg] will require all `$yes` matchers specified and must also negate
+        // all previous matchers.
+        #[cfg(all(
+            $($yes,)?
+            not(any($($no),*))
+        ))]
+        cfg_match! { @__identity $($tokens)* }
+
+        // Recurse to emit all other items in `$rest`, and when we do so add all
+        // our `$yes` matchers to the list of `$no` matchers as future emissions
+        // will have to negate everything we just matched as well.
+        cfg_match! {
+            @__items ($($no,)* $($yes,)?);
+            $($rest,)*
+        }
+    },
+
+    // Internal macro to make __apply work out right for different match types,
+    // because of how macros match/expand stuff.
+    (@__identity $($tokens:tt)*) => {
+        $($tokens)*
+    }
```

Comment:
Let's take this opportunity to simplify the implementations significantly:

```suggestion
pub macro cfg_match {
    (
        _ => { $($output:tt)* }
    ) => {
        $($output)*
    };
    
    (
        $cfg:meta => $output:tt
        $($( $rest:tt )+)?
    ) => {
        #[cfg($cfg)]
        cfg_match! {
            _ => $output
        }
        
        $(
            #[cfg(not($cfg))]
            cfg_match! { $($rest)* }
        )?
    }
```

Reply:
Wow, you are always impressing me, @danielhenrymantilla! Thank you very much!

The suggested implementation is indeed really elegant.

```rust
cfg_match! {
    windows => { let _ = 1; }
    feature = "foo" => { let _ = 2; }
    unix => { let _ = 3; }
    _ => { let _ = 4; }
}
```

The current algorithm roughly expands the above snippet into a combination of different configuration targets.

```rust
#[cfg(all(windows, not(any(feature = "foo", unix))))]
cfg_match! { @__identity let _ = 1; }

#[cfg(all(feature = "foo", not(any(windows, unix))))]
cfg_match! { @__identity let _ = 2; }

#[cfg(all(unix, not(any(windows, feature = "foo"))))]
cfg_match! { @__identity let _ = 3; }

#[cfg(all(not(any(windows, feature = "foo", unix))))]
cfg_match! { @__identity let _ = 4; }
```

But now everything expands into a nested exclusionary structure.

```rust
cfg_match! {
    #[cfg(windows)]
    let _ = 1;
    
    #[cfg(not(windows))]
    cfg_match! {
        #[cfg(feature = "foo")]
        let _ = 2;
    
        #[cfg(not(feature = "foo"))]
        cfg_match! {
            #[cfg(unix)]
            let _ = 3;
        
            #[cfg(not(unix))]
            cfg_match! {
                let _ = 4;
            }
        }
    }
}
```

I am just leaving the comma (`,`) and parentheses `=> ( ... )` matters on hold until more individuals express their options about possible further inclusions or modifications regarding the syntax, if any.

Reply:
:+1: for the change, conceptually. Either way we're going to want to teach rustdoc some way to understand it and present the alternatives, but that can come later.

Reply:
Looks like that's it then! Let's move forward with everything.